# PHGv2 - Imputation

In this document, we will discuss imputation strategies using the
PHG.

_WIP_

## hVCF export

>[!NOTE]
> This currently is needed, but will be removed in the future as we can do direct connections to the hVCF TileDB instance.

First we need to export the hVCF data from the TileDB instance. This is done using the `export-vcf` command.

```shell
./phg export-vcf --db-path /my/db/uri --dataset-type hvcf --sample-Names LineA,LineB --output-dir /my/hvcf/dir
```
This command takes in 4 parameters:
* `--db-path` - path to directory storing the TileDB instances. The
  AGC compressed genomes will be placed here on completion.
* `--dataset-type` - the type of dataset to export. In this case, we
  are exporting the hVCF data.
* `--sample-Names` - a comma-separated list of sample names to export.
* `--output-dir` - the directory to place the exported hVCF files.

## Kmer Indexing

Once we have the hVCF data exported, we need to index the kmers. This is done using the `build-kmer-index` command.

```shell
./phg build-kmer-index --db-path /my/db/uri --hvcf-dir /my/hvcf/dir 
```

This command has 2 required parameters:
* `--db-path` - path to directory storing the TileDB instances. The
  AGC compressed genomes are also stored here as well under db-path/assemblies.agc.
* `--hvcf-dir` - the directory containing the hVCF files. This is the
  output directory from the `export-vcf` command.  Right now this is required, but will be optional in the future.

This will store the kmer index in the --hvcf-dir directory named kmerIndex.txt. 

## Read Mapping

Now that we have the index we can align short reads against the PHG using the `map-kmers` command.

```shell  
./phg map-kmers --hvcf-dir /my/hvcf/dir --kmer-index /my/hvcf/dir/kmerIndex.txt --read-files /path/to/reads/LineA_R1.fq /path/to/reads/LineA_R2.fq --output-dir /my/mapping/dir
```

This command has the following parameters:
* `--hvcf-dir` - the directory containing the hVCF files.
* `--kmer-index` - the kmer index file created by the `build-kmer-index` command.
* `--read-files` - a comma separated list of fastq files for a single sample.  Either 1(for single end) or 2(for paired end) files can be input at a time this way.  Any more and an error will be thrown.
* `--output-dir` - the directory to place the read mapping files.

Instead of using `--read-files`, you can use `--key-file` to specify a key file that contains a list of the read mapping files. 
Columns for samplename and filename are required.  If using paired end fastqs, a filename2 column can be included




## Find Paths

FindPaths is the command used to impute paths through a HaplotypeGraph based on a set of read mappings.
The method uses read mapping files generated by the map-kmers command. Running FindPaths uses the Viterbi algorithm to solve a
hidden Markov model (HMM) to identify the set of haplotypes most likely to have generated the observed read mappings. A haploid
path is a single path through the graph. A diploid path is two paths through the graph that jointly explain the read mappings.

### Likely Ancestors
The term "ancestors" refers to the taxa/assemblies used to construct the haplotype graph used for imputation. The term "ancestors" is used because the model
considers the haplotypes in the graph to represent the potential ancestral haplotypes from which the samples
to be imputed were derived. In some scenarios, limiting the number of ancestors used to impute a sample will be useful.
For example, if some samples were derived from an F1 cross and the individuals used to build the graph include
those parents, using only those parents for imputation can improve accuracy. Also, because imputing diploid paths
is more computationally intensive than haploid paths, limiting the number of ancestors used per sample may be necessary to
control the amount of time required per sample.

To restrict the number of ancestors used, set use-likely-ancestors to true, and provide a value for either max-ancestors
or min-coverage. For the case where the samples have been derived from a limited number of ancestors setting
min-coverage to 0.95 or max-ancestors to the expected number of ancestors is a useful strategy. In either case, providing
a name for the output file saves a record of the ancestors used for each sample and should be checked to make sure the samples
behaved as expected.

When using this method, ancestors are chosen by first counting the number of reads for a sample that map to each ancestor in the haplotype graph.
The ancestor with the most mapped reads is chosen. Next the reads not mapping to first chosen ancestor are used to count reads mapping
to the remaining ancestors and the ancestor with the most reads is selected. This is repeated until either the proportion
of reads mapping to the selected ancestors >= min-coverage or the number ancestors selected equals max-ancestors. If a
name is provided for the likely-ancestor-file, the chosen ancestors are written to that file in the order they were chosen
along with the cumulative coverage, which is the proportion of reads mapping to each ancestor and the previously chosen ones.

### Required Parameters

**path-keyfile**: the path to a keyfile that contains a list of the read mapping files. The keyfile must have two columns
labeled SampleName and ReadMappingFiles. Any additional columns will be ignored. SampleName must be unique.
Additionally if an output hvcf for a sample name already exists in the output directory, the sample will be skipped.
If a ReadMappingFiles value is a comma-separated list of file names, the read mappings will be combined.
The file names must include the full path to the file.

**hvcf-dir**: The directory containing the hvcf used to build the haplotype graph used for imputation.

**reference-genome**: The name and path to the reference genome fasta or fastq file.

**output-dir**: The directory where the output hvcfs will be written. One file will be written for each sample, and the output file name will be (sampleName).h.vcf.

**path-type**: The type of path to be imputed. The value must be either "haploid" or "diploid" (without quotes).

### Optional Parameters

**prob-correct**: The probability that a mapped read was mapped correctly. (Default = 0.99)

**prob-same-gamete**: The probability of transitioning to the same gamete (sample) in the next reference range.
This should be equal to 1 - (probability of a recombination). Probability of a recombination can be estimated as the
total number of expected recombinations in a sample divided by the number of reference ranges (Default = 0.99)

**min-gametes**: The minimum number of gametes with a haplotype in a reference range. Reference ranges with fewer gametes will not be imputed. (Default = 1)

**min-reads**: The minimum number of reads per ReferenceRange. Reference ranges with fewer reads will not be imputed.
If minReads = 0, all ReferenceRanges will be imputed. (Default = 0)

**inbreeding-coefficient**: The estimated coefficient of inbreeding for the samples being evaluated. Only used for diploid path type. The value must be between 0.0 and 1.0 (Default = 0.0)

**max-reads-per-kb**: ReferenceRanges with more than max-reads-per-kb will not be imputed. (Default = 1000)

**use-likely-ancestors**: The value must be "true" or "false" (no parentheses). This indicates whether the most
likely ancestors of each sample will be used for path finding. (Default = false)

**max-ancestors**: If use-likely-ancestors = true, use at most max-ancestors. (Default = Integer.MAX_VALUE)

**min-coverage**: If use-likely-ancestors = true, use the fewest number of ancestors that together have this proportion
of mappable reads. The values must be between 0.5 and 1.0 (Default = 1.0)

**likely-ancestor-file**: If useLikelyAncestors is true, a record of the ancestors used for each sample will be written
to this file, if a filename is provided. (Default = "")

**threads**: number of threads used to find paths. (Default = 3)


The command
````
phg find-paths --help
````  
will print the following:  
Usage: find-paths [\<options>]

Impute best path(s) using read mappings.

Options:
--path-keyfile=\<text>          tab-delimited file with first two columns: SampleName, ReadMappingFiles. 
ReadMappingFiles must be the full path to a read mapping file or a comma separated list of file paths. 
All sample names must be unique. Required parameter.  
--hvcf-dir=\<text>              The directory containing the hvcf files used to build a HaplotypeGraph for path finding. Required parameter.  
--reference-genome=\<text>      path to reference genome (fasta or fastq). Required parameter.  
--output-dir=\<text>            The directory where the output hvcfs will be written. The output file names will be<sampleName>.h.vcf. Required parameter.  
--path-type=(haploid|diploid)  The type of path to find. Must be lower case 'haploid' or 'diploid' (without quotes). 
'haploid' infers a single path through the graph. 'diploid' infers a pair of paths. Required parameter.  
--prob-correct=\<float>         The probability that a mapped read was mapped correctly. Default = 0.99  
--prob-same-gamete=\<float>     The probability of transitioning to the same gamete (sample) in the next reference range. Default = 0.99  
--min-gametes=\<int>            The minimum number of gametes with a haplotype in a reference range. Reference ranges with fewer gametes will not be imputed. Default = 1  
--min-reads=\<int>              The minimum number of reads per ReferenceRange. Reference ranges with fewer reads will not be imputed. 
If minReads = 0, all ReferenceRanges will be imputed. Default = 0  
--inbreeding-coefficient=\<float>The estimated coefficient of inbreeding for the samples being evaluated. Only used for diploid path type. Default = 0.0  
--max-reads-per-kb=\<int>       ReferenceRanges with more than max-reads-per-kb will not be imputed. Default = 1000.  
--use-likely-ancestors=true|false  Use only the most likely ancestors of each sample for path finding. Default = false  
--max-ancestors=\<int>          If use-likely-ancestors = true, use at most max-ancestors. Default = Int.MAX_VALUE.  
--min-coverage=\<float>         If use-likely-ancestors = true, use the fewest number of ancestors that together have this proportion of mappable reads. Default = 1.0  
--likely-ancestor-file=\<text>  If useLikelyAncestors is true, a record of the ancestors used for each sample will be written to this file. Default = ''  
--threads=\<int>                number of threads used to find paths. Default = 3.  
-h, --help                     Show this message and exit  
